client/blocks/conversations/list.jsx

```diff
@@ -94,34 +95,102 @@ export class ConversationCommentList extends React.Component {
 		const { haveEarlierCommentsToFetch, haveLaterCommentsToFetch } = props.commentsFetchingStatus;
 
 		if ( enableCaterpillar && ( haveEarlierCommentsToFetch || haveLaterCommentsToFetch ) ) {
-			props.requestPostComments( { siteId, postId } );
+			const direction = haveEarlierCommentsToFetch ? 'before' : 'after';
+			props.requestPostComments( { siteId, postId, direction } );
 		}
 	};
 
 	componentDidMount() {
 		this.reqMoreComments();
 	}
 
+	componentWillReceiveProps( nextProps ) {
+		const { hiddenComments, commentsTree, siteId } = nextProps;
+
+		// if we are running low on comments to expand then fetch more
+		if ( size( hiddenComments ) < FETCH_NEW_COMMENTS_THRESHOLD ) {
+			this.reqMoreComments();
+		}
+
+		// if we are missing any comments in the hierarchy towards a comment that should be shown,
+		// then load them one at a time. This is not the most efficient method, ideally we could
+		// load a subtree
+		const inaccessible = this.getInaccessibleParentsIds(
+			commentsTree,
+			Object.keys( this.getCommentsToShow() )
+		);
+		inaccessible.forEach( commentId => {
+			nextProps.requestComment( {
+				commentId,
+				siteId,
+			} );
+		} );
+	}
+
+	getParentId = ( commentsTree, childId ) =>
+		get( commentsTree, [ childId, 'data', 'parent', 'ID' ] );
+	commentHasParent = ( commentsTree, childId ) => !! this.getParentId( commentsTree, childId );
+	commentParentIsLoaded = ( commentsTree, childId ) =>
+		get( commentsTree, this.getParentId( commentsTree, childId ), false );
+
+	getInaccessibleParentsIds = ( commentsTree, commentIds ) => {
+		// base case
+		if ( size( commentIds ) === 0 ) {
+			return [];
+		}
+
+		const withParents = filter( commentIds, id => this.commentHasParent( commentsTree, id ) );
+		const accessible = map(
+			filter( withParents, id => this.commentParentIsLoaded( commentsTree, id ) ),
+			id => this.getParentId( commentsTree, id )
+		);
+		const inaccessible = map(
+			filter( withParents, id => ! this.commentParentIsLoaded( commentsTree, id ) ),
+			id => this.getParentId( commentsTree, id )
+		);
+
+		return inaccessible.concat( this.getInaccessibleParentsIds( commentsTree, accessible ) );
+	};
+
+	getCommentsToShow = () => {
```

Comment:
wonder if we should do this work as part of willMount / willReceiveProps? doing it every render might get spendy? probably nothing to worry about now though.